### PR TITLE
Add `#[derive(FromRef)]`

### PR DESCRIPTION
**Pull request by davidpdrsn**
Given how straight forward this is I think it makes sense to support.

## TODO

- [x] Docs
- [x] Tests

### CODE REVIEW

File: axum-macros/src/from_ref.rs [hunk 1]
```diff
@@ -0,0 +1,47 @@
+use proc_macro2::{Ident, TokenStream};
+use quote::{quote, quote_spanned};
+use syn::{spanned::Spanned, Field, ItemStruct};
+
+pub(crate) fn expand(item: ItemStruct) -> syn::Result<TokenStream> {
```
Comment (jplatte) [inline axum-macros/src/from_ref.rs:5]:
```suggestion
pub(crate) fn expand(item: ItemStruct) -> TokenStream {
```

**Comment by davidpdrsn** [comment]
Yes I think so! Will add that later!

File: axum-core/src/extract/from_ref.rs [hunk 2]
```diff
@@ -5,7 +5,10 @@
 ///
 /// See [`State`] for more details on how library authors should use this trait.
 ///
+/// This trait can be derived using `#[derive(axum_macros::FromRef)]`.
+///
 /// [`State`]: https://docs.rs/axum/0.6/axum/extract/struct.State.html
+/// [`#[derive(axum_macros::FromRef)]`]: https://docs.rs/axum-macros/latest/axum_macros/derive.FromRef.html
```
[the remark this answers is not in the document]
Comment (davidpdrsn) [reply axum-core/src/extract/from_ref.rs:11]: Weren't there something with docsrs not building dev dependencies, and that is why we have the `__private_docs` feature in axum?

File: axum-macros/src/lib.rs [hunk 3]
```diff
@@ -575,6 +576,47 @@ pub fn derive_typed_path(input: TokenStream) -> TokenStream {
...
+///
+/// #
+/// # type AuthToken = String;
+/// # type DatabasePool = ();
+/// #
+/// // This will implement `FromRef` for each field in the struct.
+/// #[derive(FromRef)]
+/// struct AppState {
+///     auth_token: AuthToken,
+///     database_pool: DatabasePool,
```
[the remark this answers is not in the document]
Comment (davidpdrsn) [reply axum-macros/src/lib.rs:595]: Yeah I've gone back and forth on this. We can add that later if someone requests it.

File: axum-core/src/extract/from_ref.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (jplatte) [reply axum-core/src/extract/from_ref.rs:11]: Hm yeah, that might be right.

**Pull request merged**